packages/utils-debugging-protocol-common/src/RequestResponse.ts

```diff
@@ -0,0 +1,271 @@
+import { atob } from 'abab';
+import { Crdp } from 'chrome-remote-debug-protocol';
+
+import { getContentTypeData} from 'hint/dist/src/lib/utils/content-type';
+import { HttpHeaders, Response } from 'hint/dist/src/lib/types';
+import { debug as d } from 'hint/dist/src/lib/utils/debug';
+import { Requester } from '@hint/utils-connector-tools/dist/src/requester';
+import { normalizeHeaders } from '@hint/utils-connector-tools/dist/src/normalize-headers';
+
+import { AsyncHTMLElement } from './cdp-async-html';
+
+const debug: debug.IDebugger = d(__filename);
+
+export enum RequestStatus {
+    willBeSent = 'willBeSent',
+    responseReceived = 'responseReceived',
+    loadingFinished = 'loadingFinished',
+    loadingFailed = 'loadingFailed'
+}
+
+export class RequestResponse {
+    private _overrideInvalidCert: boolean;
+
+    /** The debugging protocol Network Client used to download the response body */
+    private _network: Crdp.NetworkClient;
+
+    private _requestId: string;
+    /** The associated `requestId` for this network request. */
+    public get requestId() {
+        return this._requestId;
+    }
+
+    private _status: RequestStatus;
+    /** The current status of the request. */
+    public get status() {
+        return this._status;
+    }
+
+    private _hops: string[] = [];
+    /** The redirects done to get to the final resource. */
+    public get hops() {
+        return this._hops;
+    }
+
+    private _originalUrl: string;
+    /** The initial URL that started the request process before any redirects. */
+    public get originalUrl() {
+        return this._originalUrl;
+    }
+
+    /** The final URL after redirects. Could be the same as `originalUrl`. */
+    public get finalUrl(): string {
+        // `willBeSent` gets updated with each redirect so this is the final requested url
+        return this.willBeSent.request.url;
+    }
+
+    private _willBeSent: Crdp.Network.RequestWillBeSentEvent | undefined;
+    /** The payload returned by the debugging protocol in the `RequestWillBeSent` event. */
+    public get willBeSent(): Crdp.Network.RequestWillBeSentEvent {
+        return this._willBeSent!;
```

Comment:
Is there a reason this one is assumed not to be `undefined` whereas the others may be?

Reply:
Nope, left over from when I was doing tests.